##Goals
- Create a feedforward neural network for classifying CIFAR10 dataset
- Various activation functions and their pros/cons
- Use tensor board for visualiazing

In [11]:
# imports and setup
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import torch.nn.functional as F
import logging
import matplotlib.pyplot as plt
import numpy as np
from torch.utils.tensorboard import SummaryWriter
from sklearn.model_selection import train_test_split

# Set up logging
logging.basicConfig(level=logging.INFO)
writer = SummaryWriter('runs/cifar10_experiment_11')

# Set random seed for pytorch and numpy
# numpy seed takes care of numpy and scipy
torch.manual_seed(42)
np.random.seed(42)

# Check if GPU is available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
logging.info(f"Using {device} device")

# classes from CIFAR10
classes = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

INFO:root:Using cpu device


In [12]:
# Data loading and preprocessing
def imshow(ax, img):
    npimg = img.numpy()
    # normalize the image from 0 to 1, cifar10 has it from -1 to 1
    npimg = (npimg+1)/2
    # imshow expects color channel to be the third dimension
    # and it expects the RGB values to be between 
    ax.imshow(np.transpose(npimg, (1, 2, 0)))

def visualize_training_examples(train_dataset, classes, num_images=6):
    # Select num_images random indices
    indices = np.random.choice(len(train_dataset), size=num_images, replace=False)

    # Show images and labels
    # figure size should be dependent on 
    plt.figure(figsize=(8, (num_images//2)*4))
    for i, idx in enumerate(indices):
        ax = plt.subplot(num_images//2, 2, i + 1)
        image, label = train_dataset[idx]
        imshow(ax, image)
        ax.set_title(classes[label])

    plt.tight_layout()
    plt.show()


def load_data(dataset_loc='./data',batch_size=64):
    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.5,), (0.5,))  # Adjust these values for normalization
    ])

    # Dataset class stores features and target
    # DataLoader builds an iterator on top of Dataset class

    # Load CIFAR10 dataset
    train_dataset = datasets.CIFAR10(root=dataset_loc, train=True, download=True, transform=transform)
    test_dataset = datasets.CIFAR10(root=dataset_loc, train=False, download=True, transform=transform)

    # Use stratified sampling to split the train dataset into train and validation
    train_dataset, validation_dataset = train_test_split(train_dataset, test_size=0.2, random_state=42, stratify=train_dataset.targets)

    # Log some info about the dataset type and size
    logging.info(f"Train dataset size: {len(train_dataset)}")
    logging.info(f"Test dataset size: {len(test_dataset)}")

    # visualize_training_examples(train_dataset, classes, num_images=20)

    # Data loaders
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    validation_loader = DataLoader(validation_dataset, batch_size=batch_size, shuffle=True)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

    return train_loader, validation_loader, test_loader

In [13]:
class FullyConnectedNN(nn.Module):
    def __init__(self):
        super(FullyConnectedNN, self).__init__()
        self.fc1 = nn.Linear(3 * 32 * 32, 2048)
        self.bn1 = nn.BatchNorm1d(2048)
        self.fc2 = nn.Linear(2048, 1024)
        self.bn2 = nn.BatchNorm1d(1024)
        self.fc3 = nn.Linear(1024, 512)
        self.bn3 = nn.BatchNorm1d(512)
        self.fc4 = nn.Linear(512, 256)
        self.bn4 = nn.BatchNorm1d(256)
        self.fc5 = nn.Linear(256, 128)
        self.bn5 = nn.BatchNorm1d(128)
        self.fc6 = nn.Linear(128, 10)

    def forward(self, x):
        x = x.view(-1, 3 * 32 * 32)  # Flatten the images
        x = F.relu(self.bn1(self.fc1(x)))
        x = F.relu(self.bn2(self.fc2(x)))
        x = F.relu(self.bn3(self.fc3(x)))
        x = F.relu(self.bn4(self.fc4(x)))
        x = F.relu(self.bn5(self.fc5(x)))
        x = self.fc6(x)
        return x

In [14]:
def train(model, device, train_loader, optimizer, epoch):
    model.train()
    running_loss = 0.0
    running_accuracy = 0.0
    for batch_idx, (data, target) in enumerate(train_loader):
        # move data to device for every iteration
        # gpu memory is limited, so we can't move all data at once
        data, target = data.to(device), target.to(device)
        # zero the gradients, otherwise they will accumulate
        optimizer.zero_grad()
        output = model(data)
        loss = F.cross_entropy(output, target)
        accuracy = (output.argmax(dim=1) == target).float().mean()
        # compute gradients
        loss.backward()
        # update weights
        optimizer.step()

        running_loss += loss.item()
        running_accuracy += accuracy.item()
        if (batch_idx+1) % 100 == 0: # +1 so that we don't print for 0th batch
            # Print running loss and running accuracy every 100 batches, also print fraction of epoch completed
            logging.info(f'Epoch: {epoch} [{batch_idx * len(data)}/{len(train_loader.dataset)} ({100. * batch_idx / len(train_loader):.0f}%)]\tLoss: {running_loss/100:.6f}\tAccuracy: {running_accuracy/100:.6f}')
            # logging.info(f'Epoch: {epoch} [{batch_idx * len(data)}/{len(train_loader.dataset)} ({100. * batch_idx / len(train_loader):.0f}%)]\tLoss: {running_loss:.6f}')
            # Log running loss to tensorboard
            writer.add_scalar('training loss', running_loss / 100, epoch * len(train_loader) + batch_idx)
            # Log accuracy to tensorboard
            writer.add_scalar('training accuracy', running_accuracy / 100, epoch * len(train_loader) + batch_idx)
            # writer.add_scalar('training accuracy', accuracy, epoch * len(train_loader) + batch_idx)
            running_loss = 0.0
            running_accuracy = 0.0

def test(model, device, test_loader, epoch=None, validation=False):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            # reduction='sum' means that we will get the sum of the loss instead of the mean
            # item() gives the scalar value of the loss
            test_loss += F.cross_entropy(output, target, reduction='sum').item()
            # get the index of the max log-probability
            # dim=1 means that we will get the max value for each row
            pred = output.argmax(dim=1)
            correct += pred.eq(target).sum().item()

    test_loss /= len(test_loader.dataset)

    if validation:
        logging.info(f'Validation set: Average loss: {test_loss:.4f}, Accuracy: {correct}/{len(test_loader.dataset)} ({100. * correct / len(test_loader.dataset):.0f}%)')
        writer.add_scalar('validation loss', test_loss, epoch)
        writer.add_scalar('validation accuracy', correct / len(test_loader.dataset), epoch)
    else:
        # log test loss and accuracy to tensorboard
        logging.info(f'Test set: Average loss: {test_loss:.4f}, Accuracy: {correct}/{len(test_loader.dataset)} ({100. * correct / len(test_loader.dataset):.0f}%)')
        writer.add_scalar('test loss', test_loss)
        writer.add_scalar('test accuracy', correct / len(test_loader.dataset))

In [15]:
# Main execution
batch_size = 64
dataset_loc = '/Users/saip/My Drive/machine-learning-fundamentals/datasets'
train_loader, validation_loader, test_loader = load_data(dataset_loc, batch_size)
model = FullyConnectedNN().to(device)
optimizer = optim.Adam(model.parameters(), lr=0.001, weight_decay=0.001)

for epoch in range(0,20):  # 10 epochs
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, validation_loader, epoch, validation=True)

test(model, device, test_loader)
writer.close()


Files already downloaded and verified
Files already downloaded and verified


INFO:root:Train dataset size: 40000
INFO:root:Test dataset size: 10000
INFO:root:Epoch: 0 [6336/40000 (16%)]	Loss: 1.931427	Accuracy: 0.303750
INFO:root:Epoch: 0 [12736/40000 (32%)]	Loss: 1.771827	Accuracy: 0.366406
INFO:root:Epoch: 0 [19136/40000 (48%)]	Loss: 1.723358	Accuracy: 0.373437
INFO:root:Epoch: 0 [25536/40000 (64%)]	Loss: 1.742861	Accuracy: 0.362344
INFO:root:Epoch: 0 [31936/40000 (80%)]	Loss: 1.701176	Accuracy: 0.394219
INFO:root:Epoch: 0 [38336/40000 (96%)]	Loss: 1.689433	Accuracy: 0.398281
INFO:root:Validation set: Average loss: 1.6510, Accuracy: 4113/10000 (41%)
INFO:root:Epoch: 1 [6336/40000 (16%)]	Loss: 1.670840	Accuracy: 0.398750
INFO:root:Epoch: 1 [12736/40000 (32%)]	Loss: 1.651841	Accuracy: 0.410781
INFO:root:Epoch: 1 [19136/40000 (48%)]	Loss: 1.659517	Accuracy: 0.406406
INFO:root:Epoch: 1 [25536/40000 (64%)]	Loss: 1.661509	Accuracy: 0.402187
INFO:root:Epoch: 1 [31936/40000 (80%)]	Loss: 1.661809	Accuracy: 0.409062
INFO:root:Epoch: 1 [38336/40000 (96%)]	Loss: 1.643589

In [16]:
writer = SummaryWriter('runs/cifar10_experiment_11')
for epoch in range(20,40):  # 10 epochs
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, validation_loader, epoch, validation=True)

test(model, device, test_loader)
writer.close()

INFO:root:Epoch: 20 [6336/40000 (16%)]	Loss: 1.190246	Accuracy: 0.573125
INFO:root:Epoch: 20 [12736/40000 (32%)]	Loss: 1.183392	Accuracy: 0.581094
INFO:root:Epoch: 20 [19136/40000 (48%)]	Loss: 1.208495	Accuracy: 0.569375
INFO:root:Epoch: 20 [25536/40000 (64%)]	Loss: 1.240068	Accuracy: 0.557031
INFO:root:Epoch: 20 [31936/40000 (80%)]	Loss: 1.193498	Accuracy: 0.582344
INFO:root:Epoch: 20 [38336/40000 (96%)]	Loss: 1.245668	Accuracy: 0.565625
INFO:root:Validation set: Average loss: 1.3892, Accuracy: 5134/10000 (51%)
INFO:root:Epoch: 21 [6336/40000 (16%)]	Loss: 1.167523	Accuracy: 0.587812
INFO:root:Epoch: 21 [12736/40000 (32%)]	Loss: 1.206297	Accuracy: 0.573438
INFO:root:Epoch: 21 [19136/40000 (48%)]	Loss: 1.229763	Accuracy: 0.562813
INFO:root:Epoch: 21 [25536/40000 (64%)]	Loss: 1.213844	Accuracy: 0.570312
INFO:root:Epoch: 21 [31936/40000 (80%)]	Loss: 1.218654	Accuracy: 0.566719
INFO:root:Epoch: 21 [38336/40000 (96%)]	Loss: 1.251702	Accuracy: 0.559063
INFO:root:Validation set: Average loss:

v1 - 2 layers, lr=0.001,batch_size=64

- Training accuracy for the model kept increasing till 90% but validation accuracy is just 50%. The state of the art stands at 99.5%
- Similar trend with loss
- The model is trying to learn the noise to decrease the training loss. 
- Does this mean the model is complex enough since the training accuracy reached 90%?

v2 - 2 layers, lr=0.1, batch_size=64
- Training accuracy and validationa accuracy stand ~40%, worst performance than before.
- Maybe the lr is too high and the model is oscillating missing the local minima

v3 - 2 layers, lr=0.1, batch_size=256
- Increasing the batch_size did not have any effect. 

v4 - 4 layers, lr=0.001, batch_size=64
- Did not have any effect. 


v5 - 4 layers, lr=0.01, batch_size=64
- Worst performance. 15% accuracy in both training and validation

v6 - 4 layers, lr=0.001, batch_size=64, batch_norm
- Not a significant improvement, validation is still ~56%

v6 - 4 layers, lr=0.005, batch_size=64, batch_norm
- Effect of learning rate probably is minimal after applying batch norm

v7 - 6 layers, lr=0.001, batch_size=64, batch_norm
- Not a significant improvement. Validaiton accuracy stands at ~56% while training accuracy reaches 80%

v7 - 6 layers, lr=0.001, batch_size=64, batch_norm, regularization
- Same fate

How do I know this is the maximum accuracy that I can get with this model? 
